tests/LoggerPluginTest.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+namespace Http\Client\Common\Plugin;
+
+use Http\Client\Exception\HttpException;
+use Http\Client\Exception\NetworkException;
+use Http\Message\Formatter\SimpleFormatter;
+use Http\Promise\FulfilledPromise;
+use Http\Promise\RejectedPromise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use PHPUnit\Framework\TestCase;
+use Psr\Http\Message\UriInterface;
+
+final class LoggerPluginTest extends TestCase
+{
+    private TestLogger $logger;
+    private LoggerPlugin $plugin;
+
+    protected function setUp(): void
+    {
+        $this->logger = new TestLogger();
+        $this->plugin = new LoggerPlugin($this->logger, new SimpleFormatter());
+    }
+
+    public function testLogsRequestAndResponse()
+    {
+        $request = $this->createMock(RequestInterface::class);
+        $response = $this->createMock(ResponseInterface::class);
+
+        $uri = $this->createMock(UriInterface::class);
+        $uri->expects($this->any())->method('__toString')->willReturn('http://example.com/');
```

Comment:
i am a bit worried about the amount of mocking here. could we require-dev something like https://github.com/Nyholm/psr7 and use concrete classes rather than mock what is essentially value objects?

with request/response its a bit of an edge case i guess, as they have some complexity. 

but at the end of the day, if things don't work with Nyholm/psr7, we need to either bugfix that repository or its actually a bug on our side.

Reply:
Yes please! To be honest, I was horrified by this mocking as well. I rarely mock in my day to day. But didn't know if we could pull in dependencies. Will refactor it.

Reply:
Much better 😍